File: toomcook.py
#toomcook
import time

def split_number(n, m):
    n_str = str(n)
    x0 = int(n_str[-m:])
    x1 = int(n_str[-2*m:-m]) if len(n_str) > m else 0
    x2 = int(n_str[:-2*m]) if len(n_str) > 2*m else 0
    return x0, x1, x2

def toom_cook_mul(x, y):
    if x < 10 or y < 10:
        return x * y

    # Calculating size of the numbers
    m = max(len(str(x)), len(str(y))) // 2

    # Spliting the numbers into three parts
    x0, x1, x2 = split_number(x, m)
    y0, y1, y2 = split_number(y, m)

    # Evaluating the polynomials at 0, 1, and -1
    x_evaluations = [x0, x0 + x1 + x2, x0 - x1 + x2]
    y_evaluations = [y0, y0 + y1 + y2, y0 - y1 + y2]

    # Pointwise multiplications
    r = [toom_cook_mul(x_eval, y_eval) for x_eval, y_eval in zip(x_evaluations, y_evaluations)]

    # Interpolating results (reconstruction)
    result = r[0] + ((r[2] - r[0] - r[1]) // 2) * 10**m + r[1] * 10**(2*m)
    return result

if __name__ == "__main__":
    n = int(input("Enter the first number: "))
    m = int(input("Enter the second number: "))
    start_time = time.time()
    print(f"The result of {n} multiplied by {m} is {toom_cook_mul(n, m)}")
    end_time = time.time()
    execution_time = end_time - start_time
    print(f"Execution Time: {execution_time} seconds")

File: toom cook.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>

#define MAX_SIZE 1024

// Function prototypes
long long toom_cook_mul(long long x, long long y);
void split_number(long long n, int m, long long *parts);

void split_number(long long n, int m, long long *parts) {
    for (int i = 0; i < 3; i++) {
        parts[i] = n % (long long)pow(10, m);
        n /= (long long)pow(10, m);
    }
}

long long toom_cook_mul(long long x, long long y) {
    if (x < 10 || y < 10) {
        return x * y;
    }

    int m = MAX_SIZE / 6;
    long long x_parts[3], y_parts[3];
    split_number(x, m, x_parts);
    split_number(y, m, y_parts);

    long long p0 = x_parts[0] * y_parts[0];
    long long p1 = (x_parts[0] + x_parts[1] + x_parts[2]) * (y_parts[0] + y_parts[1] + y_parts[2]);
    long long p2 = (x_parts[0] - x_parts[1] + x_parts[2]) * (y_parts[0] - y_parts[1] + y_parts[2]);

    long long result = p0 + ((p2 - p0 - p1) / 2) * (long long)pow(10, m) + p1 * (long long)pow(10, 2 * m);
    return result;
}

int main() {
    long long n, m;
    printf("Enter the first number: ");
    scanf("%lld", &n);
    printf("Enter the second number: ");
    scanf("%lld", &m);

    clock_t start_time = clock();
    long long result = toom_cook_mul(n, m);
    clock_t end_time = clock();

    printf("The result of %lld multiplied by %lld is %lld\n", n, m, result);
    printf("Execution Time: %lf seconds\n", (double)(end_time - start_time) / CLOCKS_PER_SEC);

    return 0;
}
